kyo-core/shared/src/main/scala/kyo/layers.scala

```diff
@@ -0,0 +1,91 @@
+package kyo
+
+import Layers.internal.*
+import kyo.core.*
+import scala.annotation.targetName
+
+sealed trait Layer[+Out, -S]:
+    self =>
+
+    infix def to[Out2, S2, In2](that: Layer[Out2, Envs[In2] & S2]): Layer[Out2, S & S2]          = To(self, that)
+    infix def and[Out2, S2](that: Layer[Out2, S2]): Layer[Out & Out2, S & S2]                    = And(self, that)
+    infix def using[Out2, S2, In2](that: Layer[Out2, Envs[In2] & S2]): Layer[Out & Out2, S & S2] = self and (self to that)
+
+    private[kyo] def doRun(
+        memoMap: scala.collection.mutable.Map[Layer[?, ?], Any] = scala.collection.mutable.Map.empty[Layer[?, ?], Any]
+    ): TypeMap[Out] < S =
+        type Expected = TypeMap[Out] < S
+        memoMap.get(self) match
+            case Some(result) => result.asInstanceOf[Expected]
+            case None =>
+                self match
+                    case And(lhs, rhs) =>
+                        {
+                            for
+                                leftResult  <- lhs.doRun(memoMap)
+                                rightResult <- rhs.doRun(memoMap)
+                            yield leftResult.union(rightResult)
+                        }.asInstanceOf[Expected]
+
+                    case To(lhs, rhs) =>
+                        {
+                            for
+                                leftResult  <- lhs.doRun(memoMap)
+                                rightResult <- Envs.runTypeMap(leftResult)(rhs.doRun(memoMap))(using summon, Envs.bypass)
+                            yield rightResult
+                        }.asInstanceOf[Expected]
+
+                    case FromKyo(kyo) =>
+                        kyo().map { result =>
+                            memoMap += (self -> result)
+                            result
+                        }.asInstanceOf[Expected]
+        end match
+    end doRun
+
+end Layer
+
+object Layer:
+    import Envs.HasEnvs
+
+    extension [In, Out, S](layer: Layer[Out, Envs[In] & S])
+        def run[In1, R](using HasEnvs[In1, In] { type Remainder = R }): TypeMap[Out] < (S & R) =
+            layer.doRun().asInstanceOf[TypeMap[Out] < (S & R)]
+end Layer
+
+object Layers extends KyoApp:
+    private[kyo] object internal:
+        case class And[Out1, Out2, S1, S2](lhs: Layer[Out1, S1], rhs: Layer[Out2, S2]) extends Layer[Out1 & Out2, S1 & S2]
+        case class To[Out1, Out2, S1, S2](lhs: Layer[?, ?], rhs: Layer[?, ?])          extends Layer[Out1 & Out2, S1 & S2]
+        case class FromKyo[In, Out, S](kyo: () => TypeMap[Out] < (Envs[In] & S))(using val tag: Tag[Out]) extends Layer[Out, S]
+    end internal
+
+    val empty: Layer[Any, Any] = FromKyo { () => TypeMap.empty }
+
+    def apply[A: Tag, S](kyo: => A < S): Layer[A, S] =
+        FromKyo { () =>
+            kyo.map { result => TypeMap(result) }
+        }
```

Comment:
I'm very happy with the result here! Without the `Envs[B]` specialization, the layer should be transparent to the underlying Kyo computation.